src/Servers/Kestrel/Core/src/Internal/TlsConnectionFeature.cs

```diff
@@ -14,29 +15,111 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal
 
+        public TlsConnectionFeature(SslStream sslStream)
+        {
+            if (sslStream is null)
+            {
+                throw new ArgumentNullException(nameof(sslStream));
+            }
+
+            _sslStream = sslStream;
+        }
+
+        public X509Certificate2? ClientCertificate
+        {
+            get
+            {
+                if (_clientCert is null)
+                {
+                    _clientCert = ConvertToX509Certificate2(_sslStream.RemoteCertificate);
+                }
+
+                return _clientCert;
+            }
+            set => _clientCert = value;
+        }
+
+        // Used for event source, not part of any of the feature interfaces.
         public string? HostName { get; set; }
 
-        public ReadOnlyMemory<byte> ApplicationProtocol { get; set; }
+        public ReadOnlyMemory<byte> ApplicationProtocol
+        {
+            get => _applicationProtocol ?? _sslStream.NegotiatedApplicationProtocol.Protocol;
+            set => _applicationProtocol = value;
+        }
 
-        public SslProtocols Protocol { get; set; }
+        public SslProtocols Protocol
+        {
+            get => _protocol ?? _sslStream.SslProtocol;
+            set => _protocol = value;
+        }
 
-        public CipherAlgorithmType CipherAlgorithm { get; set; }
+        // We don't store the values for these because they could be changed by a renegotiation.
+        public CipherAlgorithmType CipherAlgorithm
+        {
+            get => _cipherAlgorithm ?? _sslStream.CipherAlgorithm;
+            set => _cipherAlgorithm = value;
+        }
 
-        public int CipherStrength { get; set; }
+        public int CipherStrength
+        {
+            get => _cipherStrength ?? _sslStream.CipherStrength;
+            set => _cipherStrength = value;
+        }
 
-        public HashAlgorithmType HashAlgorithm { get; set; }
+        public HashAlgorithmType HashAlgorithm
+        {
+            get => _hashAlgorithm ?? _sslStream.HashAlgorithm;
+            set => _hashAlgorithm = value;
+        }
 
-        public int HashStrength { get; set; }
+        public int HashStrength
+        {
+            get => _hashStrength ?? _sslStream.HashStrength;
+            set => _hashStrength = value;
+        }
 
-        public ExchangeAlgorithmType KeyExchangeAlgorithm { get; set; }
+        public ExchangeAlgorithmType KeyExchangeAlgorithm
+        {
+            get => _keyExchangeAlgorithm ?? _sslStream.KeyExchangeAlgorithm;
+            set => _keyExchangeAlgorithm = value;
+        }
 
-        public int KeyExchangeStrength { get; set; }
+        public int KeyExchangeStrength
+        {
+            get => _keyExchangeStrength ?? _sslStream.KeyExchangeStrength;
+            set => _keyExchangeStrength = value;
+        }
 
         public Task<X509Certificate2?> GetClientCertificateAsync(CancellationToken cancellationToken)
         {
             return Task.FromResult(ClientCertificate);
```

Comment:
Cache the Task\<X509Certificate2\> in a field while you're at it?

Reply:
This method is going to get a lot more interesting in the next PR, let's hold off until then.